src/lua/datetime.lua

```diff
@@ -0,0 +1,611 @@
+        hms = true
+    end
+    local sec = obj.sec
+    if sec ~= nil then
+        check_range(sec, 0, 60, 'sec')
+        hms = true
+    end
+
+    local nsec, usec, msec = obj.nsec, obj.usec, obj.msec
+    local count_usec = bool2int(nsec ~= nil) + bool2int(usec ~= nil) +
+                       bool2int(msec ~= nil)
+    if count_usec > 1 then
+        error('only one of nsec, usec or msecs may defined simultaneously', 2)
+    end
+    if usec ~= nil then
+        self.nsec = usec * 1e3
+    elseif msec ~= nil then
+        self.nsec = msec * 1e6
+    elseif nsec ~= nil then
+        self.nsec = nsec
+    end
+
+    local ts = obj.timestamp
+    if ts ~= nil then
+        local sec_int, fraction
+        sec_int, fraction = math_modf(ts)
+        -- if there is one of nsec, usec, msec provided
+        -- then ignore fraction in timestamp
+        -- otherwise - use nsec, usec, or msec
+        if count_usec == 0 then
+            nsec = fraction * 1e9
+        else
+            error('only integer values allowed in timestamp '..
+                  'if nsec, usec, or msecs provided', 2)
+        end
+
+        self.epoch = sec_int
+        self.nsec = nsec
+
+        return self
+    end
+
+    local offset0 = self.tzoffset
+    local offset = obj.tzoffset
+    if offset ~= nil then
+        offset = get_timezone(offset)
+        check_range(offset, -720, 720, 'tzoffset')
+    end
+
+    if obj.tz ~= nil then
+        nyi('tz')
+    end
+
+    -- .year, .month, .day
+    if ymd then
+        datetime_ymd_update(self, y or y0, M or M0, d or d0, offset or offset0)
+    end
+
+    -- .hour, .minute, .second
+    if hms then
+        datetime_hms_update(self, h or h0, m or m0, sec or sec0, offset or offset0)
+    end
+
+    self.epoch, self.nsec = normalize_nsec(self.epoch, self.nsec)
+
+    if offset ~= nil then
+        self.tzoffset = offset
+    end
+
+    return self
+end
+
+local function datetime_strftime(self, fmt)
+    local strfmt_sz = 128
+    local buff = ffi.new('char[?]', strfmt_sz)
+    check_date(self, "datetime.strftime()")
+    check_str(fmt, "datetime.strftime()")
+    builtin.datetime_strftime(self, buff, strfmt_sz, fmt)
+    return ffi.string(buff)
+end
+
+local function datetime_format(self, fmt)
+    if fmt ~= nil then
+        return datetime_strftime(self, fmt)
+    else
+        return datetime_tostring(self)
+    end
+end
+
+ffi.metatype(datetime_t, {
+    __tostring = datetime_tostring,
+    __serialize = datetime_serialize,
+    __eq = datetime_eq,
+    __lt = datetime_lt,
+    __le = datetime_le,
+    __index = {
+        epoch = function(self) return self.epoch end,
+        timestamp = function(self) return self.epoch + self.nsec / 1e9 end,
+
+        usec = function(self) return self.nsec / 1e3 end,
+        msec = function(self) return self.nsec / 1e6 end,
+
+        dt = function(self) return local_dt(self) end,
+        year = function(self) return builtin.tnt_dt_year(local_dt(self)) end,
```

Comment:
Not requesting to change it now, but just this simple member access makes 2 FFI C calls + a lot of calculations in the Lua land. Seems like a very high price. In case it is not jitted, it is going to cost several hundreds of microseconds AFAIU. Quite a lot to just get a year. Such places at least could be done as a single FFI call which does all in compiled C code.

Reply:
I understand performance concerns, but these functions are not on critical path today - parsing functions are (which are in one of a followup commits, we should go review sooner). Will revisit it eventually, but not today.